src/node/hiding.rs

```diff
@@ -0,0 +1,226 @@
+use crate::node::{
+    CoreConstructible, DisconnectConstructible, JetConstructible, WitnessConstructible,
+};
+use crate::types::{Context, Error};
+use crate::{Cmr, FailEntropy, Merkelized, Word};
+
+/// Wrapper for a node type that allows the use of hidden nodes
+/// during the construction of an expression.
+///
+/// A hidden node is useful when the CMR of a given subexpression is known,
+/// but not the expression itself.
+/// In this case, the constructor [`Hiding::hidden`] can be used.
+///
+/// Besides that,
+/// [`Hiding`] implements the [`crate::node::Constructible`] trait,
+/// which brings the usual methods for program construction.
+///
+/// ## Propagation of hidden nodes
+///
+/// Hidden nodes propagate upwards in the expression tree,
+/// towards the next `case` combinator:
+///
+/// 1. `case` with a left hidden child and a right non-hidden child becomes `assertr`.
+/// 2. `case` with a left non-hidden child and a right hidden child becomes `assertl`.
+/// 3. If `case` has two non-hidden children, then it stays `case`.
+/// 4. If `case` has two hidden children, then it itself becomes hidden.
+///
+/// If there is no `case` ancestor, then the root node becomes hidden.
+/// If the root node of a program is hidden, then the program is invalid.
+///
+/// ## Retrieval of final program
+///
+/// When the program is fully constructed,
+/// then it can be retrieved via the [`Hiding::get_node`] method.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Hiding<N> {
+    result: HidingResult<N>,
+    ctx: Context,
```

Comment:
In 60166205dcb0a46cacba79c3306a8b30505a38b8:

Where is `ctx` used?

Reply:
The context is needed for `CoreConstructible::inference_ctx`.

Reply:
Ahh, ok. So now I'm worried about type inference here. I will post a top-level comment. This might be OK but my intuition is that hidden nodes should somehow return a "dummy inference context" while non-hidden nodes can just pass through to `self.result.inference_context()`.